Add support for creating regex and subnet-based ReadFrom instances from a single string

nosan opened the pull request:
Closes #3013

Before this commit, it was not possible to use ReadFrom.valueOf for subnet and regex types. This commit introduces support for these types, as well as the use of names in underscore format

tishun commented on src/main/java/io/lettuce/core/ReadFrom.java:
```diff
@@ -193,6 +195,22 @@ public static ReadFrom valueOf(String name) {
+        if (index != -1) {
+            String type = name.substring(0, index);
+            String value = name.substring(index + 1);
+            if (LettuceStrings.isEmpty(value)) {
```
Just for the sake of posterity could we also check for index == 0 to make sure we handle something like :192.0.2.0/24?

nosan replied:
added test

tishun replied:
:) funny, I was thinking we should add another check if `type` is empty, but when I saw your test it makes much more sense to leave it as you did (because we cant assume that the `:` was put there on purpose).

Thanks, that looks perfect.

tishun commented on src/main/java/io/lettuce/core/ReadFrom.java:
```diff
@@ -193,6 +195,22 @@ public static ReadFrom valueOf(String name) {
+                return subnet(value.split(","));
+            }
+            if (type.equalsIgnoreCase("regex")) {
+                return regex(Pattern.compile(value));
```
Possibly throws `PatternSyntaxException`

If we want to avoid expanding the contract (this is a runtime exception, but assuming existing applications could be catching the `IllegalArgumentsException` it might constitute a breach of contract to them) we should perhaps wrap the  `PatternSyntaxException` inside a `IllegalArgumentsException`? What do you think?

tishun commented on src/main/java/io/lettuce/core/ReadFrom.java:
```diff
@@ -193,6 +195,22 @@ public static ReadFrom valueOf(String name) {
+                return regex(Pattern.compile(value));
+            }
+        }
+        name = name.replaceAll("_", "");
```
I think this is a good improvement.

tishun reviewed (changes requested):
Awesome, thanks for the contribution!

And thanks for the tests. I just have a few of notes.

nosan commented:
I didn't add support for the kebab-case format, but it might be worth considering. For instance, `order-sensitive-regex` is more readable than `orderSensitiveRegex`. However, as an alternative, users can opt for `ORDER_SENSITIVE_REGEX`

nosan commented:
I'm not sure that `ReadFrom.valueOf(...)` needs to support order-sensitive-regex. For me, it seems like an unnecessary overhead. For more complex configurations like this, Spring Boot already provides `LettuceClientConfigurationBuilderCustomizer`.

What do you think, @wilkinsona?

wilkinsona commented:
I'd defer to @tishun and @mp911de on that one. If use of an order-sensitive regex is at least as common as a "normal" regex, I think it would be good for `valueOf` to support them both.

mp911de commented:
> I didn't add support for the kebab-case format, but it might be worth considering. For instance, order-sensitive-regex is more readable than orderSensitiveRegex. However, as an alternative, users can opt for ORDER_SENSITIVE_REGEX

In doubt, I suggest defaulting to non-order-sensitive to simplify the problem space. Once a future requirement comes along that asks for order-sensitive regex (the same could be for order-sensitive subnet) we can still sort that issue out later.

For now, I suggest including the changes that fit in nicely and where we know that we like their design.

tishun commented:
> In doubt, I suggest defaulting to non-order-sensitive to simplify the problem space. Once a future requirement comes along that asks for order-sensitive regex (the same could be for order-sensitive subnet) we can still sort that issue out later.
> 
> For now, I suggest including the changes that fit in nicely and where we know that we like their design.

@mp911de you were discussing if we should support order-sensitive subnet and not the case here, right?
If so - I also agree we should start with the non-order-sensitive and we can extend the contract if we decide we need to.

nosan commented:
Thanks, @mp911de and @tishun

1. I replaced the underscore format with the kebab-case format.
2.  I removed support for `order-sensitive-regex`

nosan commented:
@tishun 
Still, it’s worth considering support for underscores as well, as they are more familiar and commonly used for constant names (e.g., in enums).

tishun commented:
> @tishun Still, it’s worth considering support for snake_case as well or instead of kebab-case, as it is more familiar and commonly used for constant names (e.g., in enums).
> 
> Even the initial request in Spring Boot expected the snake_case format: [spring-projects/spring-boot#42576 (comment)](https://github.com/spring-projects/spring-boot/issues/42576#issue-2576622795)

Oh ... I see. Haven't considered that.
I've asked the original author of the issue about their use case.
Not entirely sure they are describing a case where they convert between enums.

But the enum conversion still makes sense to be honest - if another enum has these values (for some reason) we would simply convert based on the name (and not have to have another string representation in the alternative enum). 

Perhaps we need to support all three then. 

I will agree to both solutions (CamelCase + kebab-case and CamelCase + kebab-case + snake_case).
If we go with the first one we can extend the contract later if needed, but it is a relatively easy change to support from the start.

nosan commented:
Added support for all three formats.

nosan commented:
@tishun 

I read this comment https://github.com/spring-projects/spring-boot/issues/42576#issuecomment-2417266349

And actually,  it makes sense. Initially, the scope of this PR was to support `regex` and `subnet` types. 

Adding additional formats like `snake_case` and` kebab-case` should not be included here at all. 

In Spring Boot it will be possible to use the following format:

`spring.data.redis.lettuce.read-from=upstreamPreferred` 
`spring.data.redis.lettuce.read-from=subnet:...`

So, I propose to revert everything that is related to format changes and just add support for `regex` and `subnet`.

tishun commented:
> @tishun
> 
> I read this comment [spring-projects/spring-boot#42576 (comment)](https://github.com/spring-projects/spring-boot/issues/42576#issuecomment-2417266349)
> 
> And actually, it makes sense. Initially, the scope of this PR was to support `regex` and `subnet` types.
> 
> Adding additional formats like `snake_case` and` kebab-case` should not be included here at all.
> 
> In Spring Boot it will be possible to use the following format:
> 
> `spring.data.redis.lettuce.read-from=upstreamPreferred` `spring.data.redis.lettuce.read-from=subnet:...`
> 
> So, I propose to revert everything that is related to format changes and just add support for `regex` and `subnet`.
> 
> Supporting order-sensitive regex has diverted me from my original focus. My apologies.
> 
> **UPDATE**: I reverted the changes related to the formats. They can always be added later if necessary.

I agree. Thanks for working on this!

nosan commented:
Thank you very much, @tishun
